### PR TITLE
Use BaseVector::prepareForReuse in PrestoSerializer

### PR DESCRIPTION
**Pull request by mbasmanova**
PrestoSerializer usage of BaseVector::reuse is an anti-pattern. Any mutating API
on a vector must be called after ensuring that the vector and any of its
component (buffer, child vector) is singly-referenced. PrestoSerializer didn't
do that. This change calls BaseVector::prepareForReuse for the top-level
RowVector.

### CODE REVIEW

File: velox/serializers/PrestoSerializer.cpp [hunk 1]
```diff
@@ -551,9 +525,9 @@ void unpackTimestampWithTimeZone(
 void readTimestampWithTimeZone(
     ByteStream* source,
     velox::memory::MemoryPool* pool,
-    VectorPtr* result) {
-  VectorPtr timestamps;
-  read<int64_t>(source, BIGINT(), pool, &timestamps, false);
+    VectorPtr& result) {
+  VectorPtr timestamps = BaseVector::create(BIGINT(), 0, pool);
```
Comment (laithsakka) [inline velox/serializers/PrestoSerializer.cpp:529]: why bigint?

File: velox/serializers/PrestoSerializer.cpp [hunk 2]
```diff
@@ -582,43 +556,22 @@ void readRowVector(
...
-  if (reused) {
-    children = &reused->children();
-  } else {
-    tempChildren.resize(numChildren);
-    children = &tempChildren;
-  }
+  auto& children = result->as<RowVector>()->children();
 
-  auto childTypes = type->as<TypeKind::ROW>().children();
+  auto childTypes = type->asRow().children();
```
Comment (laithsakka) [inline velox/serializers/PrestoSerializer.cpp:569]: auto&

File: velox/serializers/PrestoSerializer.cpp [hunk 3]
```diff
@@ -641,13 +594,14 @@ void readRowVector(
...
   // if offsets is needed, reconstruct aligned layout.
   // TODO: make it efficient
   if (needOffsets) {
+    std::vector<VectorPtr> tempChildren;
     tempChildren.resize(numChildren);
-    auto sourceRow = dynamic_cast<RowVector*>((*result).get());
+    auto sourceRow = dynamic_cast<RowVector*>(result.get());
     for (int32_t child = 0; child < numChildren; ++child) {
       tempChildren[child] = BaseVector::create(childTypes[child], size, pool);
       auto src = sourceRow->childAt(child);
```
Comment (laithsakka) [inline velox/serializers/PrestoSerializer.cpp:607]: src&

File: velox/serializers/PrestoSerializer.cpp [hunk 4]
```diff
@@ -641,13 +594,14 @@ void readRowVector(
     }
   }
 
-  readNulls(source, size, reused);
+  readNulls(source, size, *result);
 
   // if offsets is needed, reconstruct aligned layout.
```
Comment (laithsakka) [inline velox/serializers/PrestoSerializer.cpp:599]: if(! needOffsets){return;}

File: velox/serializers/PrestoSerializer.cpp [hunk 5]
```diff
@@ -2008,12 +1962,15 @@ void PrestoVectorSerde::deserialize(
...
+  if (*result && result->unique()) {
+    VELOX_CHECK((*result)->type() == type);
+    (*result)->prepareForReuse();
+  } else {
     *result = std::dynamic_pointer_cast<RowVector>(
         BaseVector::create(type, numRows, pool));
-  } else {
-    (*result)->resize(numRows);
   }
+  (*result)->resize(numRows);
```
Comment (laithsakka) [inline velox/serializers/PrestoSerializer.cpp:1973]: move this to line 1968?

**Review by laithsakka — COMMENTED** [review]
looks good shall we test it on presto cluster before landing?

**Review by laithsakka — APPROVED** [review]

**Comment by mbasmanova** [comment]
> looks good shall we test it on presto cluster before landing?

Thank you for review. Yes, we will.

File: velox/serializers/PrestoSerializer.cpp [hunk 6]
```diff
@@ -689,15 +646,15 @@ void readColumns(
...
+    std::vector<VectorPtr>& result,
     bool useLosslessTimestamp) {
   static std::unordered_map<
       TypeKind,
       std::function<void(
           ByteStream * source,
           std::shared_ptr<const Type> type,
           velox::memory::MemoryPool * pool,
-          VectorPtr * result,
+          VectorPtr & result,
```
Comment (xiaoxmeng) [inline velox/serializers/PrestoSerializer.cpp:657]: nit: Is it formatted? Thanks!

**Review by xiaoxmeng — APPROVED** [review]
@mbasmanova LGTM. Thanks!

File: velox/serializers/PrestoSerializer.cpp (hunk 6 above)
Comment (mbasmanova) [reply velox/serializers/PrestoSerializer.cpp:657]: It does look strange indeed. I assume this is a bug in auto-formatter.

File: velox/serializers/PrestoSerializer.cpp [hunk 7]
```diff
@@ -641,30 +594,34 @@ void readRowVector(
...
+  if (!needOffsets) {
+    return;
+  }
+
+  std::vector<VectorPtr> tempChildren;
+  tempChildren.resize(numChildren);
+  auto sourceRow = dynamic_cast<RowVector*>(result.get());
+  for (int32_t child = 0; child < numChildren; ++child) {
+    tempChildren[child] = BaseVector::create(childTypes[child], size, pool);
+    auto& src = sourceRow->childAt(child);
```
Comment (kgpai) [inline velox/serializers/PrestoSerializer.cpp:610]: Are we guaranteed children will be non nulL ?

File: velox/serializers/PrestoSerializer.cpp [hunk 8]
```diff
@@ -2008,11 +1968,14 @@ void PrestoVectorSerde::deserialize(
   const bool useLosslessTimestamp = prestoOptions.useLosslessTimestamp;
   auto codec = common::compressionKindToCodec(prestoOptions.compressionKind);
   auto numRows = source->read<int32_t>();
-  if (!(*result) || !result->unique() || (*result)->type() != type) {
+
+  if (*result && result->unique()) {
+    VELOX_CHECK((*result)->type() == type);
+    (*result)->prepareForReuse();
```
Comment (kgpai) [inline velox/serializers/PrestoSerializer.cpp:1974]: nit: Would it be lot of additional work to change deserialize api to take shared_ptr<RowVector>& result ?

**Review by kgpai — COMMENTED** [review]
LGTM.

File: velox/serializers/PrestoSerializer.cpp (hunk 8 above)
Comment (mbasmanova) [reply velox/serializers/PrestoSerializer.cpp:1974]: This is doable, but probably not in this PR.

File: velox/serializers/PrestoSerializer.cpp (hunk 7 above)
Comment (mbasmanova) [reply velox/serializers/PrestoSerializer.cpp:610]: I think so.

**Pull request closed without merging**